protocol/src/main/java/org/mifos/chatbot/protocol/slack/SlackChatService.java

```diff
@@ -61,56 +72,115 @@ public void connect(ChatCallBack callback) {
+                        }
+                    }
+                    return;
                 }
-
-                if(callback != null && authenticated) {
-                    Message m = new Message();
-                    m.setFrom(event.getSender().getUserMail());
-                    m.setText(event.getMessageContent());
-
-                    callback.onMessage(m);
-
-                    log.info("Slack : " + event.getMessageContent());
-                    List<MifosResponse> responseList = adapterService.handle(event.getMessageContent());
-
-                    if(!responseList.isEmpty()) {
-                        for (MifosResponse response : responseList) {
-                            callback.onResponse(response);
+                User user = userRepository.findUserBySlackID(senderId);
+                if(user != null) {
+                    String username = user.getUsername();
+                    String password = user.getSecret_Pass();
+                    if(authUser(username, password)) {
+                        ApiClient apiClient = new ApiClient(base64Encode(username + ":" + password));
+                        org.mifos.chatbot.client.Configuration.setDefaultApiClient(apiClient);
+                        List<MifosResponse> responseList = adapterService.handle(messageText.toLowerCase());
+                        if (!responseList.isEmpty()) {
+                            for (MifosResponse response : responseList) {
+                                sendTextMessage(senderId, response.getContent());
+                            }
+                        } else {
+                            sendTextMessage(senderId, "Sorry i didn't get that.");
                         }
-                    } else if(!event.getMessageContent().equals(noIntent)) {
-                        MifosResponse errorResponse = new MifosResponse();
-                        errorResponse.setContent(noIntent);
-                        callback.onResponse(errorResponse);
+                    } else {
+                        sendTextMessage(senderId, "Your previous credentials are not correct. Please login again.");
+                        userRepository.removeUser(username);
                     }
+                } else {
+                    sendTextMessage(senderId, "Please login first.");
                 }
             });
         } catch (Exception e) {
             log.error(e.toString(), e);
         }
     }
 
+    private boolean authUser(String username, String password) {
+        HttpClient client = HttpClientBuilder.create().build();
+        HttpPost post = new HttpPost(settings.getApiUrl() + "/authentication?username=" + username + "&password=" + password + "&tenantIdentifier=mifos");
+        try {
+            HttpResponse response = client.execute(post);
+            if(response.getStatusLine().getStatusCode() == 200) {
+                return true;
+            }
+        } catch (Exception e) {
```

Comment:
Specific exceptions please

Reply:
same as previous, changed to IOException